File: javascript/src/median_of_two_sorted_arrays.js
/**
 * @param {number[]} nums1
 * @param {number[]} nums2
 * @return {number}
 */
function findMedianSortedArrays(nums1, nums2) {
    if (nums1.length < nums2.length) {
        return searchMedian(nums1, nums2);
    } else {
        return searchMedian(nums2, nums1);
    }
}

function searchMedian(left, right) {
    const mid = Math.floor((left.length + right.length + 1) / 2);
    let lo = 0;
    let hi = left.length;
    while (lo <= hi) {
        let mid1 = Math.floor((lo + hi) / 2);
        let mid2 = mid - mid1;

        let maxLeftLeft = mid1 === 0 ? -Infinity : left[mid1 - 1];
        let minRightLeft = mid1 === left.length ? Infinity : left[mid1];

        let maxLeftRight = mid2 === 0 ? -Infinity :  right[mid2 - 1];
        let minRightRight = mid2 === right.length ? Infinity : right[mid2];

        if (maxLeftLeft <= minRightRight && maxLeftRight <= minRightLeft) {
            if ((left.length + right.length) % 2 === 0) {
                return (
                    Math.max(maxLeftLeft, maxLeftRight)
                    + Math.min(minRightLeft, minRightRight)
                ) / 2;
            } else {
                return Math.max(maxLeftLeft, maxLeftRight);
            }
        } else if (maxLeftLeft > minRightRight) {
            hi = mid1 - 1;
        } else {
            lo = mid1 + 1;
        }
    }
}

export {findMedianSortedArrays};

File: elixir/lib/median_of_two_sorted_arrays.ex
defmodule MedianOfTwoSortedArrays do
  @moduledoc """
  See: <https://leetcode.com/problems/median-of-two-sorted-arrays/>
  """

  def find_median_sorted_arrays(nums1, nums2) do
    find_median({nums1, nums2}, {nums1, nums2}, 0)
  end

  defp find_median({[], []}, mid, index) when rem(index, 2) == 1 do
    min_head(mid)
  end

  defp find_median({[], []}, mid, index) when rem(index, 2) == 0 do
    (min_head(mid) + min_head(pop_min(mid))) / 2
  end

  defp find_median(values, mid, index) when index == 0 or rem(index, 2) == 1 do
    find_median(pop_min(values), mid, index + 1)
  end

  defp find_median(values, mid, index) when rem(index, 2) == 0 do
    find_median(pop_min(values), pop_min(mid), index + 1)
  end

  defp pop_min({[_ | left], []}), do: {left, []}
  defp pop_min({[], [_ | right]}), do: {[], right}

  defp pop_min({left, right}) do
    if hd(left) < hd(right) do
      {tl(left), right}
    else
      {left, tl(right)}
    end
  end

  defp min_head({[left | _], []}), do: left
  defp min_head({[], [right | _]}), do: right
  defp min_head({[left | _], [right | _]}), do: min(left, right)
end
